### PR TITLE
Allow content resolvers to declare their output path

### PR DESCRIPTION
**Pull request by Prokyonn**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | yes
| BC breaks? | yes
| Deprecations? | no
| Fixed tickets | -
| Related issues/PRs | supersedes #9045
| License | MIT
| Documentation PR | sulu/sulu-docs# 

#### What's in this PR?

`ResolverInterface` gains two methods: `getType()` keys the resolver's output and `getOutputPath()` places it in the resolved content view data. `getOutputPath()` is a bracket path relative to the root. Returning `null` keeps the location used today, `[extension][<type>]`, so a resolver that does not care says nothing more than its type. The empty string is the opposite and merges the output into the root itself. A path ending in `content` also writes the view to the sibling `view` key. Core resolvers keep the location they have today: `[content]` for `template`, `''` for `settings` and `null` for the rest.

Both were tag attributes in the first draft of this PR. They now live on the interface as instance methods, which means the container no longer needs to derive anything at build time: implementations are tagged by `registerForAutoconfiguration()`, the `sulu_content.content_resolver` tag carries nothing, and `ContentViewResolver` keys the output by `$resolver->getType()` at runtime. The normalizer builds its type to path map from the same tagged iterator, so the map and the keys cannot disagree. The compiler pass the earlier draft used is deleted.

Resolvers may share a path. The tag's `priority` orders them and the first writer of a key keeps it, so on a collision the higher priority resolver wins, whatever the colliding values are. There is no emptiness special case, so a resolver can return `null` and mean it. The envelope keys `content`, `view` and `extension` are seeded before any resolver runs, so a root resolver cannot claim them at any priority.

The only runtime failure left is a resolver on the root returning a non-array, which throws because there is no key to put the value under.

#### Why?

Today everything a resolver returns lands under `extension.<type>`, except `template` and `settings`, which are hardcoded in the normalizer. That fits `seo` and `excerpt`, but not resolvers that describe what the resource *is* (a product, an event): templates end up reading `extension.product.code`. #9045 addressed this with a `placement: root` flag and a hand-maintained list of reserved keys; this PR makes the output location a property of the resolver instead of a third special case.

Reading the type off the interface rather than the tag is what allows autoconfiguration. It also removes a class of silent bug: Symfony resolves a tagged iterator's index at build time, where `#[AsTaggedItem]` is only honoured for autoconfigured definitions and otherwise falls back to the service id, so a compiler pass rebuilding the same map by hand can key it differently from the iterator and drop the configured path without an error.

#### Example Usage

```php
// flat bag: {{ product.code }}
final class ProductResolver implements ResolverInterface
{
    public function resolve(DimensionContentInterface $dimensionContent, ?array $properties = null): ?ContentView
    {
        return ContentView::create(['code' => '...'], []);
    }

    public function getType(): string
    {
        return 'product';
    }

    public function getOutputPath(): ?string
    {
        return '[product]'; // '[product][content]' to also get {{ product.view.code }}
    }
}
```

```php
// nothing to register when the bundle's services are autoconfigured
$services->set('acme_product.product_resolver', ProductResolver::class)
    ->tag('sulu_content.content_resolver'); // only when autoconfigure is disabled
```

```twig
{{ product.code }}  {# instead of extension.product.code #}
```

#### To Do

- [ ] Create a documentation PR
- [x] Add upgrade notes to UPGRADE-3.x.md

### CODE REVIEW

File: packages/content/config/resolvers.php [hunk 1]
```diff
@@ -99,10 +99,10 @@
             new Reference('.inner'),
             new Reference('sulu_admin.form_metadata_provider'),
         ])
-        ->tag('sulu_content.content_resolver', ['type' => 'template']);
+        ->tag('sulu_content.content_resolver', ['type' => 'template', 'path' => '[root][content]']);
```
Comment (Prokyonn) [inline packages/content/config/resolvers.php]: content resolver should also only be on root

File: packages/content/src/Application/ContentResolver/DataNormalizer/ContentViewDataNormalizer.php [hunk 2]
```diff
@@ -43,53 +56,116 @@ public function __construct(
...
+                    /** @var array<string, mixed> $typeView */
+                    $typeView = $view[$type] ?? [];
+                    $this->mergeAt($result, [...\array_slice($segments, 0, -1), 'view'], $typeView, $type);
+                }
+            } catch (ResolverPlacementException $e) {
+                if ($this->debug) {
+                    throw $e;
+                }
 
+                $this->logger?->error('Skipped content resolver "{type}": {message}', [
```
Comment (Prokyonn) [inline packages/content/src/Application/ContentResolver/DataNormalizer/ContentViewDataNormalizer.php]: not sure how we should handle this, but crashes on resolving are never good 😬

File: packages/content/config/resolvers.php [hunk 3]
```diff
@@ -99,10 +104,10 @@
...
 
     $services->set('sulu_content.settings_resolver', SettingsResolver::class)
-        ->tag('sulu_content.content_resolver', ['type' => 'settings']);
+        ->tag('sulu_content.content_resolver', ['type' => 'settings', 'path' => '[root]']);
```
Comment (alexander-schranz) [inline packages/content/config/resolvers.php]: I'm unsure about Tag options here, it does conflicts with todays autoconfigure possibilities of Symfony.

The question is how many will build custom resolvers in the future. Aslong as we not introduce `registerForAutoconfiguration` for content resolvers we are save. To keep the tags here over a seperate or extended interface.

Another possibility would be require PHP 8.3 with default methods on interface where we could add a new method, or we decide todo a break here as I think there are not lot of new resolvers around.

File: packages/content/src/Application/ContentResolver/DataNormalizer/ContentViewDataNormalizer.php [hunk 4]
```diff
@@ -13,25 +13,38 @@
...
+    /**
+     * @param array<string, list<string>> $paths resolver type to path segments without the
+     *                                           `[root]` anchor, set by ContentResolverPathPass
+     */
     public function __construct(
-        private PropertyAccessorInterface $propertyAccessor
+        private PropertyAccessorInterface $propertyAccessor,
+        private array $paths,
+        private ?LoggerInterface $logger = null,
+        private bool $debug = false,
```
Comment (alexander-schranz) [inline packages/content/src/Application/ContentResolver/DataNormalizer/ContentViewDataNormalizer.php]: Debug should be part of a decorating service, middleware or visitor. Not part of the service itself.

File: packages/content/src/Application/ContentResolver/DataNormalizer/ContentViewDataNormalizer.php [hunk 5]
```diff
@@ -43,53 +56,140 @@ public function __construct(
...
 
+            try {
+                $result = $this->mergeResolverOutput($result, $segments, $data, $type);
+
+                if ([] !== $segments && 'content' === $segments[\count($segments) - 1]) {
+                    /** @var array<string, mixed> $typeView */
+                    $typeView = $view[$type] ?? [];
+                    $result = $this->mergeResolverOutput($result, [...\array_slice($segments, 0, -1), 'view'], $typeView, $type);
+                }
+            } catch (InvalidResolverOutputException $e) {
```
Comment (alexander-schranz) [inline packages/content/src/Application/ContentResolver/DataNormalizer/ContentViewDataNormalizer.php]: Would not introduce a `InvalidResolverOutputException` else we should define a better interface or value object if we can not work with specific outputs or handle things as they are. Seems like it was also not needed before?

File: packages/content/src/Application/ContentResolver/DataNormalizer/ContentViewDataNormalizer.php [hunk 6]
```diff
@@ -13,25 +13,38 @@
...
+
+    /**
+     * @param array<string, list<string>> $paths resolver type to path segments without the
+     *                                           `[root]` anchor, set by ContentResolverPathPass
+     */
     public function __construct(
-        private PropertyAccessorInterface $propertyAccessor
+        private PropertyAccessorInterface $propertyAccessor,
+        private array $paths,
+        private ?LoggerInterface $logger = null,
```
Comment (alexander-schranz) [inline packages/content/src/Application/ContentResolver/DataNormalizer/ContentViewDataNormalizer.php]: would not log it error should be thrown out with full stack trace.

Our property resolvers should handle invalid JSON data already as empty content.

File: packages/content/src/Application/ContentResolver/DataNormalizer/ContentViewDataNormalizer.php (hunk 2 above)
Comment (alexander-schranz) [reply packages/content/src/Application/ContentResolver/DataNormalizer/ContentViewDataNormalizer.php]: yes, but would not catch exception here. it indicates a bug here on our side, and we should not silence bugs.

If the data is changed or different because content manager switched something ( template, block types, ... ), its an expected behaviour and so even not an error, and so not needs a log, like we do for property resolvers if the return does not match.

File: packages/content/src/Application/ContentResolver/DataNormalizer/ContentViewDataNormalizer.php [hunk 7]
```diff
@@ -43,53 +56,140 @@ public function __construct(
...
+        $existing = $target[$segment];
+
+        if ([] === $rest) {
+            if (!\is_array($existing) || !\is_array($data)) {
+                throw new InvalidResolverOutputException($type, $walked, \sprintf(
+                    'that path already holds %s and cannot take %s',
+                    \get_debug_type($existing),
+                    \get_debug_type($data),
+                ));
+            }
```
Comment (alexander-schranz) [inline packages/content/src/Application/ContentResolver/DataNormalizer/ContentViewDataNormalizer.php]: I think we should just decide if first or later wins in such situations.

File: packages/content/src/Application/ContentResolver/DataNormalizer/ContentViewDataNormalizer.php [hunk 8]
```diff
@@ -43,53 +56,140 @@ public function __construct(
...
+
+            $target[$segment] = $existing + $data;
+
+            return $target;
+        }
+
+        if (!\is_array($existing)) {
+            throw new InvalidResolverOutputException($type, $walked, \sprintf('that path already holds %s', \get_debug_type($existing)));
```
Comment (alexander-schranz) [inline packages/content/src/Application/ContentResolver/DataNormalizer/ContentViewDataNormalizer.php]: Same priorty of content resolver, should matter here, what wins over what. To skip merging in such case.

**Review by alexander-schranz — CHANGES_REQUESTED** [review]
In deep its hard for me to review and follow.

General design question is that I would avoid `InvalidResolverOutputException`. At the end I think, merge aslong as possible, else we should define priority does later wins over first or first over later values and how are the deep merged. What about `null`, `''`, `[]` handling in merging.

**Comment by Prokyonn** [comment]
> What about `null`, `''`, `[]` handling in merging.

We can either say those are valid values and it will be used as value or even overwrite existing content or filter them out completely. I would be more in favor of keeping them, to not have any more magic edits in the code.

**Comment by alexander-schranz** [comment]
> What about null, '', [] handling in merging.
> > We can either say those are valid values and it will be used as value or even overwrite existing content or filter them out completely. I would be more in favor of keeping them, to not have any more magic edits in the code.

I would not filter out completely we just need think if a `'title' => 'Test'` is not be overwritten / preferred over `'title' => null` or `'title' => ''` during merge.

File: packages/content/src/Application/ContentResolver/Resolver/ResolverInterface.php [hunk 9]
```diff
@@ -26,4 +36,16 @@ interface ResolverInterface
      * @param array<string, string>|null $properties
      */
     public function resolve(DimensionContentInterface $dimensionContent, ?array $properties = null): ?ContentView;
+
+    /**
+     * Keys this resolver's output, e.g. `seo`.
+     */
+    public static function getType(): string;
```
Comment (Prokyonn) [inline packages/content/src/Application/ContentResolver/Resolver/ResolverInterface.php]: do we require the static?

File: packages/content/src/Application/ContentResolver/Resolver/ResolverInterface.php [hunk 10]
```diff
@@ -26,4 +36,16 @@ interface ResolverInterface
...
+
+    /**
+     * Where the output lands in the resolved content view data, as a bracket path anchored at
+     * `[root]`, e.g. `[root][extension][seo]`. A path ending in `content` also writes the view
+     * to the sibling `view` key.
+     */
+    public static function getOutputPath(): string;
```
Comment (Prokyonn) [inline packages/content/src/Application/ContentResolver/Resolver/ResolverInterface.php]: same as above

File: packages/content/src/Infrastructure/Symfony/HttpKernel/Compiler/ContentResolverPathPass.php [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+/**
+ * Reads `getOutputPath()` off every `sulu_content.content_resolver` and writes the result into the
+ * normalizer's `$paths` argument, a map from resolver type to path segments without the `[root]`
+ * anchor. Two resolvers may share a path; the normalizer merges them by priority.
+ * Runs after `DecoratorServicePass`, so a decorated service already carries its decorator's class
+ * and both are read as one resolver.
+ *
+ * @internal this class is not part of the public API and should only be called by the Symfony framework classes
+ */
+class ContentResolverPathPass implements CompilerPassInterface
```
Comment (Prokyonn) [inline packages/content/src/Infrastructure/Symfony/HttpKernel/Compiler/ContentResolverPathPass.php]: can be removed

**Review by alexander-schranz — APPROVED** [review]
As discussed looks good to me. We need check it then if we implement the product bundle if all works as expected.

Static methods as said I would avoid if possible, as decorating such service does not always work great.
